### PR TITLE
Refactor - Resource References + Constructing ARM Functions

### LINKED ISSUE #195: Refactor: Cross-Scope Resource Deployments
## Description
Implement the concept of "Resource Scopes" to the Azure Provider, to enable cross-scope deployments. Azure outlines the need for this [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/cross-scope-deployment?tabs=azure-powershell).

### What Are Resource Scopes?

Scopes are best understood by reviewing an Azure ResourceId: 

#### Azure ResourceId Example - KeyVault Secret
`"/subscriptions/0000000-0000-0000-0000-00000000000/resourceGroups/example-integration-rossmurr4y-seg-baseline/providers/Microsoft.KeyVault/vaults/vault-1234567/secrets/mgmt-baseline-secret-ConnectionKey"`

Each section in the above resourceId is a different scope that this particular resource exists in. It has:
* a subscription of `0000000-0000-0000-0000-00000000000`
* a resourceGroup of `example-integration-rossmurr4y-seg-baseline`
* a provider of `Microsoft.KeyVault`
* a root parent resource
  * of the type `vaults`
  * with a name of `vault-1234567`
* finally it defines itself
  * with a type itself of `secrets`
  * and a name of `mgmt-baseline-secret-ConnectionKey`
  
It is important to understand that each segment of the resourceId is itself a new scope for that resource. However, for the remainder of this explanation we are only going to talk about the `subscription` and `resourceGroup` scopes.

### Scopes at Deployment Time
At the time of deployment, an ARM Template can be either deployed to a "ResourceGroup" scope (indicating that all resources contained in the template are deploying to a specified resourceGroup) or to a "Deployment" scope, indicating that all resources within the arm template exist above the scope of "ResourceGroup" - resources such as an "Account" only exist at this level - there is no "ResourceGroup" deployment for an "Account". This functionality already exists within the Hamlet `manageDeployment.sh` script in the Engine.

To look at it another way, a deployment of an ARM template at the `resourceGroup` scope will typically look like this:

```sh
az login

<login to an Azure Account and Subscription> 

az deployment group create --resource-group "my-example-resource-group" --template-file ./template.json
```

This will set the scope of all resources contained within the associated template to the `subscription` scope of the logged in account, and the `resourceGroup` scope of "my-example-resource-group".

### Scopes within an ARM Template
Because of the assignment of these two scopes at deployment time, all resources within an ARM template must be capable of their assignment. However, there are at least two common ways that this may not be a possibility. 

#### The Problems
1. a Resource within an ARM template simply does not exist at the Resource Group scope, where the rest of the resources in the template do.

2. A Resource has a parent that already exists within another scope.

#### Azure's Provided Solution

Azure provide a resource specifically for this purpose called  the "Deployment Resource". This resource type effectively defines additional deployments with alternate scopes within the one ARM Template. The "Deployment Resource" has a typical resource structure like any other, however through the use of a unique `template` property, an entire ARM Template can be "nested" within the resource. 

<details>
<summary>KeyVault Secret Resource (parent in the same resourceGroup)</summary>
<br>

```json
    {
      "name": "vault-1234567/exampleSecret",
      "type": "Microsoft.KeyVault/vaults/secrets",
      "apiVersion": "2018-02-14",
      "properties": {
         ...
      },
      "dependsOn": [
        ...
      ]
    }
```

</details>

<details>
<summary>KeyVault Secret Resource (nested inside a Deployment Resource)</summary>
<br>

```json
{
      "name": "exampleDeploymentResource",
      "type": "Microsoft.Resources/deployments",
      "apiVersion": "2019-10-01",
      "properties": {
        "template": {
          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {},
          "resources": [
            {
              "name": "exampleSecret",
              "type": "Microsoft.KeyVault/vaults/secrets",
              "apiVersion": "2018-02-14",
              "properties": {
                ...
              }
            }
          ],
          "outputs": {
            "secretXfake": {
              "type": "string",
              "value": "[resourceId('Microsoft.KeyVault/vaults/secrets', 'rosstest')]"
            },
            "secretXfakeXname": {
              "type": "string",
              "value": "rosstest"
            }
          }
        }
      }
    }
```

</details>

### LINKED ISSUE #196: Refactor: Simplify GetReference
## Description
Currently the `getReference()` function implements the logic of scopes (see #195) without knowing it, and in such a way that it cannot grow with the rest of the product. Every call to `getReference()` has to be carefully constructed as the function has too many parameters. It is also doing too many things - `getReference()` should simply do one of two things:
1. obtain a reference for something in the current deployment unit >> format an ARM function accordingly.
2. if its not in the current deployment-unit, assume the resource is already deployed and perform a `getExistingReference` lookup of it.

**note**: The recent implementation of scopes complicates point 1 - referencing a resource in the current deployment-unit will first need to review the scope of that resource and format the reference to lookup the necessary scope.

## Possible Solution

- [x] Incorporate a new scope attribute into the Azure Resource Profile (done in #198 )
- [ ] refactor the way ARM Functions are constructed, so they can be as flexible as possible
- [ ] refactor `getReference()` to perform only the actions of steps 1 and 2 above (accounting for the **note**)
- [ ] Update every call to `getReference()` to the new structure.

## Expected Outcomes
* should not require further changes to `getReference()` in the Azure provider
* it will be far easier to understand and align with the usage of the AWS and Shared providers
* all ARM functions will be supported without further functions/macros - will not need to create/support individual functions
* creating a new reference will be far simpler than it is now - minimal required function parameters.

### PR DESCRIPTION
**Pull request by RossMurr4y**
Description
Implements the following changes, enabled by #200
* `@armResource` refactored to implement scopes, which will nest resources as necessary.
* Added `constructArmOutputsFromMappings` function which creates appropriately scoped outputs
* `getReference()` for the Azure provider completely reworked and drastically simplified
* Same for `getChildReference()` previously `getSubReference()`
* Standardised the approach to constructing  ARM functions - can now create raw (no boilerplate) functions and standard functions as needed, nesting them to enable more complex functionality that is occasionally required
* Completed the transition from “pseudo” resource “types” to “pseudo” as a scope
* Enabled resource type name spacing, where multiple resources under the same provider use the same name. Namespacing allows for the definition of unique names (i.e `service:apis` and `product:apis`, whilst ensuring the provider-defined type is accessible. 
	* Resource naming updated to only use the last namespace

### Additional Housekeeping Changes
* Cleaned up numerous unused functions, macros, resources and resourceProfiles
* added some missing outputs that were being implemented by other means incorrectly
* moved some discovered resource naming out of setup and into state

## Motivation and Context

Closes #195
Closes #196 

## How Has This Been Tested?

Local template generation

## Types of changes

- [ ] [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] [x] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [ ] [ ] New feature (non-breaking change which adds functionality)
- [ ] [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

- [ ] requires #200 

## Checklist:

- [ ] [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] [ ] I have updated the documentation accordingly.
- [ ] [ ] I have added tests to cover my changes.
- [ ] [ ] All new and existing tests passed.
- [ ] [x] None of the above.

### CODE REVIEW

File: azure/deploymentframeworks/arm/output.ftl [hunk 1]
```diff
@@ -37,19 +37,67 @@
...
+    [#return {
+        name : {
+            "type" : type,
+            "value" : value
+        } +
+        attributeIfContent("condition", condition)
+    }]
+[/#function]]
+
+[#function constructArmOutputsFromMappings id name scope mappings=[]]
```
Comment (ml019) [inline azure/deploymentframeworks/arm/output.ftl:50]: A few more comments please.

**Review by ml019 — CHANGES_REQUESTED** [review]
Overall this looks to clean up and simplify things.

Please ensure key core functions are well documented to assist ongoing maintenance.

File: azure/deploymentframeworks/arm/output.ftl (hunk 1 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl:57]: Have added some more, hopefully they make sense!

**Comment by roleyfoley** [comment]
@RossMurr4y Looks like we have some new conflicts, would you mind resolving them?

**Comment by RossMurr4y** [comment]
@roleyfoley  I've resolved the conflicts :)

**Pull request merged**